### PR TITLE
Fix ColladaShapes colors and orientations

### PR DESCRIPTION
**Pull request by BenjaminHug8**
**Description**
The `ColladaShapes` was not handling correctly the orientation of the meshes + it was not correctly taking into account the fact that a mesh can have multiple colors, resulting in a mix of the colors.

### CODE REVIEW

File: docs/reference/mesh.md [hunk 1]
```diff
@@ -26,6 +27,10 @@ Otherwise, the file should be specified with a relative path.
...
 Absolute paths work as well, but they are not recommended because they are not portable across systems.
 
-The `name` field defines which sub-meshe is included.
+The `name` field defines which sub-mesh is included.
 If the `name` value is an empty string then all sub-meshes are included.
 Note that the `name` field is applied only for Collada files.
+
+The `materialIndex` field defines which color of a sub-mesh is included.
+If the `materialIndex` value is strictly negative then all colors of the sub-mesh are included.
+Note that the `materialIndex` field is applied only for Collada files.
```
Comment (omichel) [inline docs/reference/mesh.md:36]: That seems clearer to me:
```suggestion
The `materialIndex` field applies only to Collada files.
It defines a material for which all the geometries associated with this material will be included.
If `materialIndex` is strictly negative then all the geometries of the Collada file are included.
```

File: resources/nodes/Mesh.wrl [hunk 2]
```diff
@@ -3,4 +3,5 @@
 Mesh {
   field MFString url [ ]   # URL of a 3D file such as DAE, STL, OBJ or FBX format (similar to ImageTexture node)
   field SFString name ""   # Filter geometries based on the given name (for example, Collada files can contain multiple geometries in a single file).
+  field SFInt32 materialIndex -1   # Filter geometries based on the given material index (for example, Collada files can contain multiple colors in a single mesh).
```
Comment (omichel) [inline resources/nodes/Mesh.wrl:6]:
```suggestion
  field SFInt32 materialIndex -1   # Filter geometries in Collada files based on the material index (Collada files can contain multiple geometries with different materials).
```

File: src/webots/nodes/WbMesh.cpp [hunk 3]
```diff
@@ -157,12 +158,47 @@ void WbMesh::updateTriangleMesh(bool issueWarnings) {
     return;
   }
 
+  if (mMaterialIndex->value() >= (int)scene->mNumMaterials) {
+    warn(tr("Geometry with the color index \"%1\" doesn't exist in the mesh.").arg(mMaterialIndex->value()));
```
Comment (omichel) [inline src/webots/nodes/WbMesh.cpp:162]:
```suggestion
    warn(tr("Geometry with color index \"%1\" doesn't exist in the mesh.").arg(mMaterialIndex->value()));
```

File: src/webots/nodes/WbMesh.cpp [hunk 4]
```diff
@@ -157,12 +158,47 @@ void WbMesh::updateTriangleMesh(bool issueWarnings) {
...
+  scene->mRootNode->mTransformation = mat;
+
   // count total number of vertices and faces
   int totalVertices = 0;
   int totalFaces = 0;
   for (unsigned int i = 0; i < scene->mNumMeshes; ++i) {
-    totalVertices += scene->mMeshes[i]->mNumVertices;
-    totalFaces += scene->mMeshes[i]->mNumFaces;
+    const aiMesh *mesh = scene->mMeshes[i];
+    if (mName->value() != "" && mName->value() != mesh->mName.data)
```
Comment (omichel) [inline src/webots/nodes/WbMesh.cpp:194]:
```suggestion
    if (!mName->value().isEmpty() && mName->value() != mesh->mName.data)
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 5]
```diff
@@ -310,14 +320,16 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
-    stream += " Transform { ";
-  stream += QString(" translation %1 %2 %3 ").arg(position[0]).arg(position[1]).arg(position[2]);
-  stream += " rotation " + webotsRotation.toString(WbPrecision::FLOAT_MAX);
-  stream += QString(" scale %1 %2 %3 ").arg(scaling[0]).arg(scaling[1]).arg(scaling[2]);
-  stream += " children [";
+  if (!referenceMeshes) {
+    if (importAsSolid)
+      stream += " Solid { ";
+    else
+      stream += " Transform { ";
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:327]: I believe we should avoid useless spaces here:
```suggestion
      stream += " Solid {";
    else
      stream += " Transform {";
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 6]
```diff
@@ -310,14 +320,16 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
+    stream += QString(" translation %1 %2 %3 ").arg(position[0]).arg(position[1]).arg(position[2]);
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:328]:
```suggestion
    stream += QString(" translation %1 %2 %3").arg(position[0]).arg(position[1]).arg(position[2]);
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 7]
```diff
@@ -310,14 +320,16 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
+    stream += " rotation " + webotsRotation.toString(WbPrecision::FLOAT_MAX);
+    stream += QString(" scale %1 %2 %3 ").arg(scaling[0]).arg(scaling[1]).arg(scaling[2]);
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:330]:
```suggestion
    stream += QString(" scale %1 %2 %3").arg(scaling[0]).arg(scaling[1]).arg(scaling[2]);
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 8]
```diff
@@ -432,22 +445,24 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
   }
 
-  if (defNeedGroup) {
-    stream += " ] ";
-    stream += " } ";
-  }
+  if (!referenceMeshes) {
+    if (defNeedGroup) {
+      stream += " ] ";
+      stream += " } ";
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:451]:
```suggestion
      stream += " ]";
      stream += " }";
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 9]
```diff
@@ -432,22 +445,24 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
+    for (unsigned int i = 0; i < node->mNumChildren; ++i)
+      addModelNode(stream, node->mChildren[i], scene, fileName, referenceFolder, importTextureCoordinates, importNormals,
+                   importAppearances, importAsSolid, importBoundingObjects, referenceMeshes);
 
-  stream += " ] ";
-  if (importAsSolid) {
-    stream += QString(" name \"%1\" ").arg(node->mName.C_Str());
-    if (importBoundingObjects && node->mNumMeshes > 0)
-      stream += " boundingObject USE SHAPE ";
+    stream += " ] ";
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:458]:
```suggestion
    stream += " ]";
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 10]
```diff
@@ -432,22 +445,24 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
+    if (importAsSolid) {
+      stream += QString(" name \"%1\" ").arg(node->mName.C_Str());
+      if (importBoundingObjects && node->mNumMeshes > 0)
+        stream += " boundingObject USE SHAPE ";
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:462]:
```suggestion
        stream += " boundingObject USE SHAPE";
```

File: src/webots/nodes/utils/WbNodeOperations.cpp [hunk 11]
```diff
@@ -432,22 +445,24 @@ static void addModelNode(QString &stream, const aiNode *node, const aiScene *sce
...
+    }
+    stream += " } ";
```
Comment (omichel) [inline src/webots/nodes/utils/WbNodeOperations.cpp:464]:
```suggestion
    stream += " }";
```

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you.
It looks good to me (only minor requests).

File: src/webots/nodes/WbMesh.cpp [hunk 12]
```diff
@@ -157,12 +158,47 @@ void WbMesh::updateTriangleMesh(bool issueWarnings) {
...
 
+  if (mMaterialIndex->value() >= (int)scene->mNumMaterials) {
+    warn(tr("Geometry with color index \"%1\" doesn't exist in the mesh.").arg(mMaterialIndex->value()));
+    return;
+  }
+
+  // Assimp fix for up_axis
+  // Adapted from https://github.com/assimp/assimp/issues/849
+  int UpAxis = 1, UpAxisSign = 1, FrontAxis = 2, FrontAxisSign = 1, CoordAxis = 0, CoordAxisSign = 1;
+  double UnitScaleFactor = 1.0;
```
Comment (omichel) [inline src/webots/nodes/WbMesh.cpp:169]: Oops, I missed this minor coding style issue: variables should be camelCase, i.e., start with a lowercase letter.

Comment (BenjaminHug8) [reply src/webots/nodes/WbMesh.cpp:169]: Thank you I will do this.

Comment (BenjaminHug8) [reply src/webots/nodes/WbMesh.cpp:169]: Fixed in https://github.com/cyberbotics/webots/pull/4193.

**Pull request merged**